analyses/npl/09_flood_events_by_municipality.md

```diff
@@ -0,0 +1,300 @@
+df_basins['Karnali'] = 0
+idx = df_basins['Major_Basi'].isin([f'{basin} River Basin' for basin in KARNALI_BASINS])
+df_basins.loc[idx, 'Karnali']= 1
+karnali = df_basins[idx].dissolve(by='Karnali')
+koshi = df_watershed[df_watershed['WSH_NME'] == f'{KOSHI_WATERSHED} Watershed']
+```
+
+Plot event municipalities over the areas of interest
+
+```python
+fig, ax = plt.subplots()
+karnali.plot(ax=ax)
+koshi.plot(ax=ax)
+df_events.boundary.plot(ax=ax, color='r', lw=0.5)
+```
+
+Might be easier to just do a latitude cut. Latitude > 86 is Koshi, < 83 is Karnali.
+
+```python
+df_events['basin'] = np.nan
+df_events.loc[df_events['geometry'].bounds.minx > 86,'basin'] = 'Koshi'
+df_events.loc[df_events['geometry'].bounds.minx < 83,'basin'] = 'Karnali'
+df_events = df_events.dropna(subset=['basin'])
+```
+
+### Combine events with GloFAS
+
+```python
+# Take only the columns with impact and group by the basin and incident date and sum
+df_events_final = df_events[df_events.columns[24:43]].groupby(['basin', 'Incident Date']).sum()
+
+```
+
+```python
+# Since there are two many events, get only those that are high impact
+# Use a rolling sum
+rolling_sum_size = 10
+center = False
+event_thresh = 0
+impact_parameters =  ['Total Death', 
+                         'Affected Family', 
+                         'Private House Partially Damaged', 
+                         'Private House Fully Damaged']
+plot_results = False
+
+df_events_high_impact = pd.DataFrame(columns=['date', 'basin', 'impact_parameter'])
+
+for basin, station in STATIONS.items():
+    # Reindex the events to the reanalysis time
+    # and get the rolling sum of impact
+    df = df_events_final.loc[basin]
+    df = (df.reindex(ds_glofas_reanalysis.time.data, fill_value=0)
+            .rolling(rolling_sum_size, center=center).sum())
+    if plot_results:
+        fig, axs = plt.subplots(len(impact_parameters), figsize=(5, 12))
+        fig.suptitle(basin)
+    for i, impact_parameter in enumerate(impact_parameters):
+        # Define event threshold as n x standard deviation
+        groups = utils.get_groups_above_threshold(df[impact_parameter], event_thresh)
+        # Take the date of the event as the first index
+        df_events_high_impact = df_events_high_impact.append(
+            [{"basin": basin,
+             "impact_parameter": impact_parameter,
+             "date": df.index[group[0]] 
+            } for group in groups],
+            ignore_index=True
+        )
+        if plot_results:
+            ax = axs[i]
+            ax.plot(df.index, df[impact_parameter], label=impact_parameter, c=f'C{i}')
+            ax.legend()
+```
+
+```python
+# Define GloFAS events
+rp = 2
+ndays = 1 # Number of consecutive days above RP
+days_before_buffer = 5 # Number of days before GloFAS event the flooding event can occur
+days_after_buffer = 60
+
+df_station_stats = pd.DataFrame(columns=['station', 'impact_parameter', 'TP', 'FP', 'FN'])
+
+for basin, station in STATIONS.items():
```

Comment:
Looks like this is repeated a bunch across notebooks so I think definitely good to generalize.

Reply:
Yes totally agree. Now that Nepal is wrapped up I actually have on my to-do list to refactor all the notebooks, in particular:
- Move all the pre-amble stuff to some kind of config file (like all the file locations etc)
- Refactor exactly this method. It's sometimes done slightly differently in some cases but I really want to try and generalize it.

I will make a new PR for this (hopefully today!)